system/HTTP/RequestTrait.php

```diff
@@ -0,0 +1,372 @@
+<?php
+
+/**
+ * This file is part of the CodeIgniter 4 framework.
+ *
+ * (c) CodeIgniter Foundation <admin@codeigniter.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace CodeIgniter\HTTP;
+
+use CodeIgniter\Validation\FormatRules;
+
+/**
+ * Request Trait
+ *
+ * Additional methods to make a PSR-7 Request class
+ * compliant with the framework's own RequestInterface.
+ *
+ * @see https://github.com/php-fig/http-message/blob/master/src/RequestInterface.php
+ */
+trait RequestTrait
+{
+	/**
+	 * IP address of the current user.
+	 *
+	 * @var string
+	 *
+	 * @deprecated Will become private in a future release
+	 */
+	protected $ipAddress = '';
+
+	/**
+	 * Stores values we've retrieved from
+	 * PHP globals.
+	 *
+	 * @var array
+	 */
+	protected $globals = [];
+
+	/**
+	 * Gets the user's IP address.
+	 *
+	 * @return string IP address
+	 */
+	public function getIPAddress(): string
+	{
+		if ($this->ipAddress)
+		{
+			return $this->ipAddress;
+		}
+
+		$ipValidator = [new FormatRules(), 'valid_ip'];
+
+		/** @deprecated $this->proxyIPs property will be removed in the future */
+		$proxyIPs = isset($this->proxyIPs) ? $this->proxyIPs : config('App')->proxyIPs;
+		if (! empty($proxyIPs) && ! is_array($proxyIPs))
+		{
+			$proxyIPs = explode(',', str_replace(' ', '', $proxyIPs));
+		}
+
+		$this->ipAddress = $this->getServer('REMOTE_ADDR');
+
+		if ($proxyIPs)
+		{
+			foreach (['HTTP_X_FORWARDED_FOR', 'HTTP_CLIENT_IP', 'HTTP_X_CLIENT_IP', 'HTTP_X_CLUSTER_CLIENT_IP'] as $header)
+			{
+				if (($spoof = $this->getServer($header)) !== null)
+				{
+					// Some proxies typically list the whole chain of IP
+					// addresses through which the client has reached us.
+					// e.g. client_ip, proxy_ip1, proxy_ip2, etc.
+					sscanf($spoof, '%[^,]', $spoof);
+
+					if (! $ipValidator($spoof))
+					{
+						$spoof = null;
+					}
+					else
+					{
+						break;
+					}
+				}
+			}
+
+			if ($spoof)
+			{
+				foreach ($proxyIPs as $proxyIP)
+				{
+					// Check if we have an IP address or a subnet
+					if (strpos($proxyIP, '/') === false)
+					{
+						// An IP address (and not a subnet) is specified.
+						// We can compare right away.
+						if ($proxyIP === $this->ipAddress)
+						{
+							$this->ipAddress = $spoof;
+							break;
+						}
+
+						continue;
+					}
+
+					// We have a subnet ... now the heavy lifting begins
+					// // @phpstan-ignore-next-line
+					isset($separator) || $separator = $ipValidator($this->ipAddress, 'ipv6') ? ':' : '.';
```

Comment:
Can we make use the normal structure instead of short circuit? I think with that it will solve the other `@phpstan-ignore-line` related to `$separator`